### PR TITLE
Reduce duplication of CompoundModel._initialize_slices

### PR DESCRIPTION
**Pull request by astrofrog**
Noticed that ``CompoundModel._initialize_slices`` was duplicating most of the cost of ``Model._initialize_slices``, so I've combined the two and all the tests seem to pass!

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by astrofrog** [comment]
Marking this as 6.1.2 because it would simplify a bug fix in #16710, but we can change to 7.0.0 if anyone has any strong feelings.

File: astropy/modeling/core.py [hunk 1]
```diff
@@ -2740,6 +2740,8 @@ def _initialize_slices(self):
             param_size = np.size(value)
             param_shape = np.shape(value)
             param_slice = slice(total_size, total_size + param_size)
+            if name not in param_metrics:
+                param_metrics[name] = {}
```
Comment (neutrinoceros) [inline astropy/modeling/core.py:2744]: This looks equivalent to
```suggestion
            param_metrics.setdefault(name, {})
```
but I also see that `param_metrics` is defined as a `defaultdict(dict)`, so I don't understand why setting a default value is even needed.

 https://github.com/astropy/astropy/blob/main/astropy/modeling/core.py#L2584

Comment (astrofrog) [reply astropy/modeling/core.py]: I dug into this further and found a place where param_metrics was set (incorrectly) to a regular dictionary, so I've fixed this.

**Review by neutrinoceros — APPROVED** [review]
LGTM now, thanks!

**Review by nden — APPROVED** [review]
LGTM

**Comment by nden** [comment]
I'm happy to merge this but am not sure what towncrier is complaining about.

**Comment by astrofrog** [comment]
I restarted and it passes

**Comment by nden** [comment]
v6.1.2 is out, which is probably why it won't let me merge. Is it OK to create a v6.1.3 milestone or should I assign 7.0.0?

**Comment by astrofrog** [comment]
I assigned 7.0.0 for now but will create 6.1.3 later. Strangely I didn't have issues merging even before changing the milestone.

**Comment by astrofrog** [comment]
I've changed it back to 6.1.2 to avoid confusion, will move it over to 6.1.3 when I do the post release steps.

**Pull request merged**